File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.*;

class Solution {
    public int[] solution(String[] operations) {
        PriorityQueue<int[]> minHeap = new PriorityQueue<>(
            (a, b) -> Integer.compare(a[0], b[0]) 
        );

        PriorityQueue<int[]> maxHeap = new PriorityQueue<>(
            (a, b) -> Integer.compare(b[0], a[0]) 
        );
        HashSet<Integer> valid = new HashSet<>();
        int idx = 0;
        
        for(String s: operations) {
            if(s.startsWith("I")) {
                int num = Integer.parseInt(s.split(" ")[1]);
                minHeap.add(new int[]{num, idx});
                maxHeap.add(new int[]{num, idx});
                valid.add(idx);
                idx += 1;
            } else if (s.equals("D 1")) {
                while (valid.size() > 0) {
                    int[] item = maxHeap.poll();
                    if (valid.contains(item[1])) {
                        maxHeap.remove(item[1]);
                        valid.remove(item[1]);
                        break;
                    }
                }
            } else if (s.equals("D -1")) {
                while (valid.size() > 0) {
                    int[] item = minHeap.poll();
                    if (valid.contains(item[1])) {
                        minHeap.remove(item[1]);
                        valid.remove(item[1]);
                        break;
                    }
                }
            }
        }
        
        if (valid.size() == 0) return new int[] {0, 0};
        
        int max_value = Integer.MIN_VALUE;
        int min_value = Integer.MAX_VALUE;
        while (valid.size() > 0) {
            int[] item = minHeap.poll();
            if (valid.contains(item[1])) {
                min_value = item[0];
                max_value = item[0];
                break;
            }
        }
        
        while (valid.size() > 0) {
            int[] item = maxHeap.poll();
            if (valid.contains(item[1])) {
                if (item[1] > max_value) max_value = item[0];
                break;
            }
        }
        return new int[]{max_value, min_value};
    }
}

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.py
import heapq

def solution(operations):
    max_heap = []
    min_heap = []
    valid = set()
    idx = 0
    
    for o in operations:
        if o.startswith("I"):
            num = int(o.split()[1])
            heapq.heappush(max_heap, (-num, idx))
            heapq.heappush(min_heap, (num, idx))
            valid.add(idx)
            idx += 1
        elif o == "D 1":
            while max_heap:
                item = heapq.heappop(max_heap)
                if item[1] in valid:
                    valid.remove(item[1])
                    break
        elif o == "D -1":
            while min_heap:
                item = heapq.heappop(min_heap)
                if item[1] in valid:
                    valid.remove(item[1])
                    break
    
    if len(valid) == 0:
        return [0, 0]
    
    min_value = 0
    max_value = 0
    while valid and min_heap:
        item = heapq.heappop(min_heap)
        if item[1] in valid:
            valid.remove(item[1])
            min_value = item[0]
            max_value = item[0]
            break
    while valid and max_heap:
        item = heapq.heappop(max_heap)
        if item[1] in valid:
            valid.remove(item[1])
            if -item[0] > max_value:
                max_value = -item[0]
            break
    return [max_value, min_value]
    